lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -4427,6 +4432,84 @@ INITIALIZE_PASS(SROA_Parameter_HLSL, "scalarrepl-param-hlsl",
   "Scalar Replacement of Aggregates HLSL (parameters)", false,
   false)
 
+void SROA_Parameter_HLSL::RewriteBitcastWithIdenticalStructs(Function *F) {
+  if (F->isDeclaration())
+    return;
+  // Gather list of bitcast involving src and dest structs with identical layout
+  std::vector<BitCastInst*> worklist;
+  for (inst_iterator I = inst_begin(F), E = inst_end(F); I != E; ++I) {
+    if (BitCastInst *BCI = dyn_cast<BitCastInst>(&*I)) {
+      Type *DstTy = BCI->getDestTy();
+      Type *SrcTy = BCI->getSrcTy();
+      if (!SrcTy->isPointerTy() || !DstTy->isPointerTy())
+        continue;
+      DstTy = DstTy->getPointerElementType();
+      SrcTy = SrcTy->getPointerElementType();
+      if (!SrcTy->isStructTy() || !DstTy->isStructTy())
+        continue;
+      StructType *DstST = cast<StructType>(DstTy);
+      StructType *SrcST = cast<StructType>(SrcTy);
+      if (!SrcST->isLayoutIdentical(DstST))
```

Comment:
For the record, as we discussed, this won't catch all the cases FXC supports (whether that's desirable or not). For FXC, struct { float2, float2} == struct { float, float2, float }. The DXC semantic layer actually also accepts this, I believe, so it will still crash here.

Reply:
Yes, FXC is far more flexible with casting not only with struct casting but in other cases too. This change would enable the scenario where the layout matches exactly. In future, we could look to extend cases like you mentioned above.